core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java

```diff
@@ -226,10 +241,10 @@ public boolean isCollectingTaskNeeded(QueryId queryId, PlanFragment plan)
      * In such case dynamic filters must be unblocked (and probe split generation resumed) for
      * source stage containing joins to allow build source tasks to flush data and complete.
      */
-    public void unblockStageDynamicFilters(QueryId queryId, PlanFragment plan)
+    public void unblockStageDynamicFilters(QueryId queryId, int attemptId, PlanFragment plan)
     {
         DynamicFilterContext context = dynamicFilterContexts.get(queryId);
-        if (context == null) {
+        if (context == null || attemptId != context.getCurrentAttemptId()) {
             // query has been removed or not registered (e.g dynamic filtering is disabled)
             return;
         }
```

Comment:
can we `checkState(attemptId == context.getCurrentAttemptId()`? So there are no attempts from the future. I assume `registerQueryRetry` will be called before any action for current attempt